### PR TITLE
[oauthlib] Improve return types of `Client` methods

### LINKED ISSUE #9907: `Client.prepare_request_uri` from `oauthlib.oauth2` does not return None
[stubs/oauthlib/oauthlib/oauth2/rfc6749/clients/base.pyi:49-51](https://github.com/python/typeshed/blob/f3fc7926207d34ae1c1e6a5166557696f5f7910f/stubs/oauthlib/oauthlib/oauth2/rfc6749/clients/base.pyi#L49) read:

```python
    def prepare_request_uri(self, *args, **kwargs) -> None: ...
    def prepare_request_body(self, *args, **kwargs) -> None: ...
    def parse_request_uri_response(self, *args, **kwargs) -> None: ...
```

Neither of these methods expected to return None.

Consider example:

```python
from oauthlib.oauth2 import WebApplicationClient

client = WebApplicationClient('foo')
uri = client.prepare_request_uri('https://foo.bar')  # This returns a string, not None as inferred
print(type(uri))
```

If we follow the spaghetti for `prepare_request_uri()` we'll eventually reach [this return statement](https://github.com/oauthlib/oauthlib/blob/7d7fe904af504c20f7d802650e54a78e3d0be8ba/oauthlib/common.py#L249) from `oauthlib`:

```
return urlparse.urlunparse((sch, net, path, par, query, fra))
```

`urlunparse` has [it's own definition](https://github.com/python/typeshed/blob/f3fc7926207d34ae1c1e6a5166557696f5f7910f/stdlib/urllib/parse.pyi#L204) in typeshed:

```python
def urlunparse(components: Sequence[AnyStr | None]) -> AnyStr: ...
```

And `AnyStr` is either bytes or str:

```python
AnyStr = TypeVar('AnyStr', bytes, str)
```

Similarly for `prepare_request_body()`, will eventually `return urlencode(queryparams)`, which doesn't specify return type:

```python
def urlencode(params): ...
```

And `parse_request_uri_response()` will eventually lead to returning `dict(urlparse.parse_qsl(query))`

So I propose to change the stubs:

```python
    def prepare_request_uri(self, *args, **kwargs) -> AnyStr: ...
    def prepare_request_body(self, *args, **kwargs) -> Any: ...
    def parse_request_uri_response(self, *args, **kwargs) -> dict: ...
```

### PR DESCRIPTION
**Pull request by aresler**
Originally reported by a user in https://youtrack.jetbrains.com/issue/PY-59638/Wrong-type-inference-for-oauthlib.oauth2.WebApplicationClient.preparerequesturi

The issue is causing false-positive for PyCharm type checker. 

Please see https://github.com/python/typeshed/issues/9907 for my motivation for these changes based on code implementation.

I'm not a typing guru, so feel free to decline if I'm missing something.

Fixes #9907

### CODE REVIEW

File: stubs/oauthlib/oauthlib/oauth2/rfc6749/clients/base.pyi [hunk 1]
```diff
@@ -46,9 +46,9 @@ class Client:
     ) -> None: ...
     @property
     def token_types(self): ...
-    def prepare_request_uri(self, *args, **kwargs) -> None: ...
-    def prepare_request_body(self, *args, **kwargs) -> None: ...
-    def parse_request_uri_response(self, *args, **kwargs) -> None: ...
+    def prepare_request_uri(self, *args, **kwargs) -> AnyStr: ...
```
Comment (srittau) [inline stubs/oauthlib/oauthlib/oauth2/rfc6749/clients/base.pyi:49]: If I read the implementation correctly, this should actually return just `str`:

```suggestion
    def prepare_request_uri(self, *args, **kwargs) -> str: ...
```

File: stubs/oauthlib/oauthlib/oauth2/rfc6749/clients/base.pyi [hunk 2]
```diff
@@ -46,9 +46,9 @@ class Client:
...
+    def prepare_request_body(self, *args, **kwargs) -> Any: ...
```
Comment (srittau) [inline stubs/oauthlib/oauthlib/oauth2/rfc6749/clients/base.pyi:50]: Same here:

```suggestion
    def prepare_request_body(self, *args, **kwargs) -> str: ...
```

File: stubs/oauthlib/oauthlib/oauth2/rfc6749/clients/base.pyi [hunk 3]
```diff
@@ -46,9 +46,9 @@ class Client:
...
+    def parse_request_uri_response(self, *args, **kwargs) -> dict[AnyStr, AnyStr]: ...
```
Comment (srittau) [inline stubs/oauthlib/oauthlib/oauth2/rfc6749/clients/base.pyi:51]:
```suggestion
    def parse_request_uri_response(self, *args, **kwargs) -> dict[str, str]: ...
```

**Review by srittau — CHANGES_REQUESTED** [review]
Thanks! Remarks below.

File: stubs/oauthlib/oauthlib/oauth2/rfc6749/clients/base.pyi (hunk 1 above)
Comment (Akuli) [reply stubs/oauthlib/oauthlib/oauth2/rfc6749/clients/base.pyi:49]: Maybe an easier way is to look at examples in the docstrings. I used `grep` to find them:

```
akuli@akuli-desktop:/tmp/e/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/clients$ grep -A1 prepare_request_uri *
base.py:    def prepare_request_uri(self, *args, **kwargs):
base.py-        """Abstract method used to create request URIs."""
--
base.py:        auth_url = self.prepare_request_uri(
base.py-            authorization_url, redirect_uri=self.redirect_url,
--
mobile_application.py:    def prepare_request_uri(self, uri, redirect_uri=None, scope=None,
mobile_application.py-                            state=None, **kwargs):
--
mobile_application.py:            >>> client.prepare_request_uri('https://example.com')
mobile_application.py-            'https://example.com?client_id=your_id&response_type=token'
mobile_application.py:            >>> client.prepare_request_uri('https://example.com', redirect_uri='https://a.b/callback')
mobile_application.py-            'https://example.com?client_id=your_id&response_type=token&redirect_uri=https%3A%2F%2Fa.b%2Fcallback'
mobile_application.py:            >>> client.prepare_request_uri('https://example.com', scope=['profile', 'pictures'])
mobile_application.py-            'https://example.com?client_id=your_id&response_type=token&scope=profile+pictures'
mobile_application.py:            >>> client.prepare_request_uri('https://example.com', foo='bar')
mobile_application.py-            'https://example.com?client_id=your_id&response_type=token&foo=bar'
grep: __pycache__: On hakemisto
--
web_application.py:    def prepare_request_uri(self, uri, redirect_uri=None, scope=None,
web_application.py-                            state=None, code_challenge=None, code_challenge_method='plain', **kwargs):
--
web_application.py:            >>> client.prepare_request_uri('https://example.com')
web_application.py-            'https://example.com?client_id=your_id&response_type=code'
web_application.py:            >>> client.prepare_request_uri('https://example.com', redirect_uri='https://a.b/callback')
web_application.py-            'https://example.com?client_id=your_id&response_type=code&redirect_uri=https%3A%2F%2Fa.b%2Fcallback'
web_application.py:            >>> client.prepare_request_uri('https://example.com', scope=['profile', 'pictures'])
web_application.py-            'https://example.com?client_id=your_id&response_type=code&scope=profile+pictures'
web_application.py:            >>> client.prepare_request_uri('https://example.com', code_challenge='kjasBS523KdkAILD2k78NdcJSk2k3KHG6')
web_application.py-            'https://example.com?client_id=your_id&response_type=code&code_challenge=kjasBS523KdkAILD2k78NdcJSk2k3KHG6'
web_application.py:            >>> client.prepare_request_uri('https://example.com', code_challenge_method='S256')
web_application.py-            'https://example.com?client_id=your_id&response_type=code&code_challenge_method=S256'
web_application.py:            >>> client.prepare_request_uri('https://example.com', foo='bar')
web_application.py-            'https://example.com?client_id=your_id&response_type=code&foo=bar'
```

If the library returns a `bytes` object instead, it is a bug in the library. Users do not need to handle `bytes` return values.

**Comment by aresler** [comment]
Sorry, forgot about this PR for a year, seem still relevant, so let's finish merging.

**Comment by srittau** [comment]
CI error is unrelated.

**Review by srittau — APPROVED** [review]

**Pull request merged**